tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala

```diff
@@ -350,6 +352,124 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
+          if (ancestorInfo ne null) {
+            ancestorInfo.methodInfos.get(proxyName) match {
+              case Some(m) =>
+                assert(m.isReflProxy && !m.isAbstract)
+                Some(m)
+
+              case _ =>
+                ancestorInfo.findProxyMatch(proxyName) match {
+                  case Some(target) =>
+                    val targetName = target.encodedName
+                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
+
+                  case None =>
+                    loop(ancestorInfo.superClass)
+                }
+            }
+          } else {
+            None
+          }
+        }
+
+        loop(this)
+      }
+    }
+
+    private def findProxyMatch(proxyName: String): Option[MethodInfo] = {
+      val candidates = methodInfos.valuesIterator.filter { m =>
+        // TODO In theory we should filter out protected methods
+        !m.isReflProxy && !m.isExported && !m.isAbstract &&
+        reflProxyMatches(m.encodedName, proxyName)
+      }.toSeq
+
+      val targets = candidates.filterNot { c =>
+        val resultType = methodResultType(c.encodedName)
+        candidates.exists { other =>
+          (other ne c) &&
+          isMoreSpecific(methodResultType(other.encodedName), resultType)
+        }
+      }
+
+      targets.headOption
+    }
+
+    private def reflProxyMatches(methodName: String, proxyName: String): Boolean = {
+      val sepPos = methodName.lastIndexOf("__")
+      sepPos >= 0 && methodName.substring(0, sepPos + 2) == proxyName
+    }
+
+    private def methodResultType(methodName: String): ir.Types.ReferenceType = {
+      val typeName = methodName.substring(methodName.lastIndexOf("__") + 2)
+      val arrayDepth = typeName.indexWhere(_ != 'A')
+      if (arrayDepth == 0)
+        ir.Types.ClassType(typeName)
+      else
+        ir.Types.ArrayType(typeName.substring(arrayDepth), arrayDepth)
+    }
+
+    private def isMoreSpecific(left: ir.Types.ReferenceType,
+        right: ir.Types.ReferenceType): Boolean = {
+      import ir.Types._
+
+      def classIsMoreSpecific(leftCls: String, rightCls: String): Boolean = {
+        leftCls != rightCls && {
+          val leftInfo = _classInfos.get(leftCls)
+          val rightInfo = _classInfos.get(rightCls)
+          leftInfo.zip(rightInfo).exists { case (l, r) =>
+            l.ancestors.contains(r)
+          }
+        }
+      }
+
+      (left, right) match {
+        case (ClassType(leftCls), ClassType(rightCls)) =>
+          classIsMoreSpecific(leftCls, rightCls)
+        case (ArrayType(leftBase, leftDepth), ArrayType(rightBase, rightDepth)) =>
+          leftDepth == rightDepth && classIsMoreSpecific(leftBase, rightBase)
+        case (ArrayType(_, _), ClassType(ObjectClass)) =>
+          true
+        case _ =>
+          false
+      }
+    }
+
+    private def createReflProxy(proxyName: String,
+        targetName: String): MethodInfo = {
```

Comment:
I forgot to mention that: Consider asserting that we are of a sensible class type (Scala classes and modules only, no traits, etc.)